llm/kv_cache.go

```diff
@@ -0,0 +1,69 @@
+package llm
+
+import (
+	"log/slog"
+	"slices"
+)
+
+// Interface for GGML functionality needed by validation
+type GGMLModel interface {
+	KV() KV
+}
+
+// kVCacheQuantization determines the appropriate KV cache quantization type based on
+// model characteristics, GPU capabilities, and environment settings
+func kVCacheQuantization(kvct string, ggml GGMLModel) string {
+	if kvct == "f16" || kvct == "f32" {
+		return kvct
+	}
+
+	// First check if the requested type is valid
+	validKVCacheTypes := []string{"f32", "f16", "q8_0", "q4_0"}
+	if !slices.Contains(validKVCacheTypes, kvct) {
+		slog.Warn("unsupported cache type, defaulting to f16", "type", kvct)
```

Comment:
We can remove the logging here as it will be done in `server.go` when determining the final runner value. It also makes this function callable in memory.go without any logging occurring twice